File: backend/model/db/fetch_topology.py

from psycopg import ServerCursor

from model.cache import cache
from model.data.device import Device
from model.data.group import Group
from model.data.link import Link
from model.device_configuration import DeviceConfiguration
from model.db.update_devices import update_topology_cache

def get_topology_as_dict():
    """
    Acquires topology from database, and converts it into a dict
    """
    devices, links, groups = cache.topology

    update_topology_cache()

    return {
        "devices": [device.to_dict() for device in devices.values()],
        "links": [link.to_dict() for link in links.values()],
        "groups": [group.to_dict() for group in groups.values()]
    }




def parse_devices(cur: ServerCursor):
    """
    Parses the output of a database call, into a dictionary of devices in dictionary form, with the device_id as the key

    If the device already exists, the data gets updated

    updates cache as result
    :return: None
    """
    devices = cache.devices
    cur.execute(
        """
            SELECT Analytics.devices.device_id, device_name, position_x, position_y, latitude, longitude, management_hostname, requested_metadata, requested_metrics, available_values 
                FROM Analytics.devices;
        """)
    for row in cur.fetchall():
        device_id = int(row[0])

        # TODO: Handle snmp configuration
        if devices.get(device_id) is None:
            devices[device_id] = Device(
                device_id=int(row[0]),
                device_name=row[1],
                position_x=float(row[2]),
                position_y=float(row[3]),
                latitude=float(row[4]),
                longitude=float(row[5]),
                management_hostname=row[6],
                configuration=DeviceConfiguration(
                    requested_metadata=set(row[7]),
                    requested_metrics=set(row[8]),
                    available_values=set(row[9]) if (row[9] is not None) else {},
                    data_sources=set(),
                )
            )

        else:
            devices[device_id].device_name = row[1]
            devices[device_id].position_x = float(row[2])
            devices[device_id].position_y = float(row[3])
            devices[device_id].latitude = float(row[4])
            devices[device_id].longitude = float(row[5])
            devices[device_id].management_hostname = row[6]
            devices[device_id].configuration = DeviceConfiguration(
                requested_metadata=set(row[7]),
                requested_metrics=set(row[8]),
                available_values=set(row[9]) if (row[9] is not None) else {},
                data_sources=set(),
            )

    # we extract the data sources, the database is normalized, so it's a separate table
    cur.execute(
        """
            SELECT device_id, data_source FROM Analytics.device_data_sources;
        """)
    for row in cur.fetchall():
        device_id = int(row[0])

        if devices.get(device_id) is not None:
            devices[device_id].configuration.data_sources.add(row[1])


def parse_device_datasource(cur: ServerCursor):
    """
    Parses the output of a database call, reading which data sources are used for each device

    If the device already exists, the data gets updated

    :return: None
    """
    devices = cache.devices
    cur.execute("SELECT device_id, data_source FROM Analytics.device_data_sources")
    for row in cur.fetchall():
        device_id = int(row[0])

        source : set = devices[device_id].configuration.data_sources
        source.add(row[1])


def parse_link(cur: ServerCursor):
    """
    Parses the output of a database call, into a dictionary of links in dictionary form, with the link id as key

    updates cache as a result
    :return: None
    """
    links = {}
    cur.execute("SELECT link_id, side_a, side_b, side_a_iface, side_b_iface, link_type, link_subtype FROM Analytics.links")
    for row in cur.fetchall():
        link = Link(
            link_id= int(row[0]),
            side_a_id = int(row[1]),
            side_b_id = int(row[2]),
            side_a_iface= row[3],
            side_b_iface= row[4],
            link_type= row[5],
            link_subtype= row[6] if row[6] is not None else ""
        )

        links[link.link_id] = link

    cache.links = links


def parse_groups(cur: ServerCursor):
    """
    Parses the output of a database call, into a list of Groups in dictionary form, with the group ID as key

    Updates cache as result
    :return: None
    """
    groups = dict[int, Group]()
    cur.execute("SELECT group_id, group_name, is_display_group FROM Analytics.groups")
    for row in cur.fetchall():
        gid = int(row[0])
        groups[gid] = Group(group_id=gid, name=row[1], members=[], is_display_group=row[2] == 't')

    group_members = {}
    cur.execute("SELECT (group_id, item_id) FROM Analytics.group_members")
    for row in cur.fetchall():
        row = row[0]
        gid = int(row[0])
        if group_members.get(gid) is None:
            group_members[gid] = []

        group_members[gid].append(int(row[1]))

    # denormalization of database
    group_list = {}
    for group in groups:
        # groups[group] # -> Group

        # if not empty, add it, brah
        if group_members.get(groups[group].group_id) is not None:
            groups[group].members = group_members[groups[group].group_id]

        group_list[groups[group].group_id] = (groups[group])

    cache.groups = group_list

    # TODO: Prevent group circular dependency

File: backend/src/model/db/fetch_topology.rs
use std::collections::{HashMap, HashSet};

use rocket::futures::TryFutureExt;
use sqlx::Postgres;
use sqlx::pool::PoolConnection;

use crate::AegisError;
use crate::misc::parse_json_array;
use crate::model::cache::Cache;
use crate::model::data::device::Device;
use crate::model::data::device_state::DeviceStatus;
use crate::model::data::link::Link;
use crate::model::data::group::Group;
use crate::model::data::device_configuration::DeviceConfiguration;
use crate::model::data::DataSource;
use crate::model::db::update_topology::update_topology_cache;
use crate::types::{DeviceId, GroupId, LinkId};

#[allow(unused)] // Needs to be allowed. Needed for compilation, but the compiler complains of a type casting needed if it's removed
use crate::model::data::link_type::LinkType;

#[derive(sqlx::FromRow)]
struct DeviceDataSource {
    device_id: DeviceId,
    fact_data_source: DataSource,
}


impl From<Option<DeviceDataSource>> for DeviceDataSource {
    fn from(value: Option<DeviceDataSource>) -> Self {
        value.unwrap_or(DeviceDataSource { device_id: 0, fact_data_source: DataSource::Icmp })
    }
}

pub async fn get_topology_as_json(pool: &sqlx::PgPool) -> Result<serde_json::Value, AegisError> {
    let mut conn = pool.acquire().await.map_err(AegisError::Sql)?;

    // Update before returning, if needed
    update_topology_cache(&mut conn, false).await?;

    // returning from Cache, Cache handles it
    Cache::instance().as_json().await
}

pub async fn get_topology_view_as_json(pool: &sqlx::Pool<Postgres>) -> Result<serde_json::Value, AegisError> {
    // Return from sql, directly into JSON because there's not need to keep a local cache or create a struct
    let rows: Vec<serde_json::Value> = sqlx::query_scalar(
        r#"
        SELECT json_build_object(
            'id', v.topology_views_id,
            'name', v.name,
            'members', COALESCE(
                json_agg(
                    json_build_object(
                        'id', m.item_id,
                        'x', m.position_x::DOUBLE PRECISION,
                        'y', m.position_y::DOUBLE PRECISION
                    )
                ) FILTER (WHERE m.item_id IS NOT NULL),
                '[]'
            )
        )
        FROM Analytics.topology_views v
        LEFT JOIN Analytics.topology_views_member m
            ON v.topology_views_id = m.topology_views_id
        GROUP BY v.topology_views_id, v.name;
        "#
    )
    .fetch_all(pool)
    .await
    .map_err(AegisError::Sql)?;

    Ok(serde_json::Value::Array(rows))
}



pub async fn query_devices(conn: &mut PoolConnection<Postgres>) -> Result<HashMap<DeviceId, Device>, AegisError>{
    // Query datasources
    let mut datasources : HashMap<DeviceId, HashSet<DataSource>> = HashMap::new();
    let rows = sqlx::query_as!(
        DeviceDataSource, 
        r#"
            SELECT device_id,
                fact_data_source::TEXT as "fact_data_source: DataSource"
            FROM Analytics.device_data_sources
            "#
    )
        .fetch_all(&mut **conn)
        .await
        .map_err(AegisError::Sql)?;

    for row in rows {
        let device_id = row.device_id;

        datasources.entry(device_id)
            .or_default()
            .insert(row.fact_data_source);
    }
    
    // Query devices themselves
    let rows = sqlx::query!(
        "SELECT Analytics.devices.device_id, device_name, latitude, longitude, management_hostname, requested_metadata, requested_metrics, available_values 
        FROM Analytics.devices;"
    )
    .fetch_all(&mut **conn)
    .await
    .map_err(|e| {
        println!("[ERROR][DB]Failed to SELECT devices from database with error = '{}'", &e.to_string());
        AegisError::Sql(e)
    })?;

    let cache = Cache::instance();

    let mut devices = HashMap::new();
    for row in rows {
        let device_id = row.device_id;

        let config = DeviceConfiguration {
            requested_metadata: parse_json_array(Some(row.requested_metadata)),
            requested_metrics: parse_json_array (Some(row.requested_metrics)),
            available_values: parse_json_array(row.available_values),
            data_sources: datasources.get(&row.device_id).unwrap_or(&HashSet::new()).clone(),
        };

        let old_device = cache.get_device(device_id).await;
        let old_state = old_device.map(|d| d.state).unwrap_or(DeviceStatus::empty());

        let device = Device {
            device_id,
            device_name: row.device_name.unwrap_or("Unnamed device".to_owned()),
            latitude: row.latitude,
            longitude: row.longitude,
            management_hostname: row.management_hostname,
            configuration: config,
            state: old_state
        };

        devices.insert(device_id, device);
    }

    Ok(devices)
}


pub async fn query_links(conn: &mut PoolConnection<Postgres>) -> Result<HashMap<LinkId, Link>, AegisError> {
    let rows = sqlx::query_as!(
        Link,
        r#"SELECT link_id, side_a, side_b, side_a_iface, side_b_iface, link_type::TEXT as "link_type: LinkType", link_subtype FROM Analytics.links;"#
    )
    .fetch_all(&mut **conn)
    .await
    .map_err(AegisError::Sql)?;

    let mut links = HashMap::new();
    for row in rows {
        links.insert(row.link_id, row);
    }

    Ok(links)
}

pub async fn query_groups(conn: &mut PoolConnection<Postgres>) -> Result<HashMap<GroupId, Group>, AegisError>{
    let rows = sqlx::query_as!(
        Group,
        r#"
            SELECT g.group_id, g.group_name as name, g.is_display_group, array_agg(gm.item_id) as members
            FROM Analytics.groups as g
            JOIN Analytics.group_members gm on gm.group_id = g.group_id
            GROUP BY g.group_id, g.group_name;
        "#
    )
    .fetch_all(&mut **conn)
    .map_err(AegisError::Sql)
    .await?;

    let mut groups = HashMap::new();

    for group in rows {
        groups.insert(group.group_id, group);
    }

    Ok(groups)
}